ADR/0039-workspace-deprecation.md

```diff
@@ -0,0 +1,98 @@
+# 39. Workspace Deprecation
+
+Some of the text was taken from - https://github.com/konflux-ci/architecture/pull/187
+
+## Status
+
+WiP
+
+## Context
+
+The purpose of this ADR is to revisit the workspace concept, understand the purpose
+of it, and offer an alternative implementation based on native Kubernetes APIs
+and other successful cloud native open source projects.
+
+We inherited the "workspace" concept from our early days when we expected to be integrated with the [KCP] project. Long ago (over a year at the time of this writing), we decided to remove [KCP] from the architecture of the Konflux project.
+
+In order to support multi-cluster deployment, [kubesaw] was used instead of KCP.
+
+Today, Workspaces are an important concept in the Konflux user experience. The user onboards into a workspace of their own. A team member of theirs may add them to a team workspace. As a platform administrator, if you hear about a user having trouble, you'll ask them what workspace they're working in.
+
+A Workspace is an abstraction of a namespace. Each workspace has a single namespace
+which backs it.
+
+Workspaces are required by [kubesaw] for providing a transparent multi-cluster communication for us. A user's workspace may be allocated to one cluster or another - but our proxy layer keeps the user blissfully unaware of this. Workspaces can be placed on clusters either automatically (using [kubesaw] mechanisms) or manually by a platform administrator.
+
+Workspace roles give a way to describe who has what rights in the workspace.
+
+Workspaces happen to be our mechanism for managing quota - a fact inherited from the [kubesaw] stack.
+
+### Some changes
+
+Some relevant changes that have occurred or are ongoing.
+
+* We want to make Konflux a successful open source projects that fits into the
+cloud native apps ecosystem. We want everything to work on vanilla Kubernetes - no OpenShift required.
+
+* We had our [first upstream Konflux release](https://github.com/konflux-ci/konflux-ci) which doesn't use [kubesaw] or relying
+on the Workspace abstraction, and is deployable on vanilla Kubernetes.
+
+* We made a decision (now long ago) to drop [KCP] from our architecture, which initially introduced the workspace concept. Both transparent multi-cluster as a feature and workspaces as a concept came from KCP.
+
+* We made the decision (at Red Hat) to stop trying to offer a managed service interface for Konflux. The user signup, waitlist, and workspace provisioning interfaces made a lot of sense there.
+
+## Decision
+
+1. We will remove the Workspace abstraction from Konflux and use standard Kubernetes namespaces
+directly for providing a place for users to run their pipelines and store their configurations
+and secrets.
+
+2. We will stop using the term Workspace, and start to use the term Namespace.
+
+3. The Konflux UI will expose a wizard for creating a new namespace. This
+wizard will be available to any user that has permissions to create namespaces.
```

Comment:
1. You can create  a validating webhook for namespaces as well.
2. Kubernetes by default doesn't limit the amount of namespaces a user can create,  I think that by default we should do the same. It would be up to the cluster admin to decide which users can create namespaces, it's not Konflux specific.

Reply:
I rather have two layers of protection, a validation webhook on workspaces not namespace and then a controller that makes sure user is the owner. For example with a CR you more options to set ownership in the status for example where it will be immutable as opposed to a label or annotation.

Reply:
@dperaza4dustbit what are the reasons you prefer the approach you are suggesting?

Reply:
Even if a validation webhook fails to spot an issue the controller will not allow non owners to do anything on controller. Ownership will be immutable if we set it in the status of the workspace. Namepsace will not be able to do that, you can do labels or maybe annotations but those can be changed. It think it just become easier to manage namespace. There is an analogy for this BTW it is called projects in OpenShift you want to provide additional values on top of the namepsace concept you can just extend the k8s API. Not sure what is the resistance to extend an API the way it was intended to be extended. We have pleany of API extensions for konflux already. This is one of the easy ones

Reply:
1. The status of a resource is mutable.

2. It should be possible to recreate the status of a resource from the state of the cluster or external state, how would you do it for the fields that holds the username? is this implementation already exist today?

3. the fact the k8s api can be extended doesn't mean it should be extended. We need to have good reasons why not to use existing resources before creating new one, otherwise we would just re invent the wheel an invest development time on thins that doesn't bring additional value.

Reply:
In addition, I don't mind that anyone would develop any extension to Konflux (including workspaces), I just don't think any extension, and specifically the proposed workspace abstraction (https://github.com/konflux-workspaces/workspaces) should be included in Konflux core, it can be an addon.

Reply:
What do we consider core btw is it everything under konflux-ci org or a subset of repos. I see workspace-manager under konflux-ci so are you suggesting we remove that as well.
Assuming you mean core is a subset of repos from the org, this happens all the time, helm organization for example has the helm repo as the helm core and many other supporting projects, like chart-museum, that enhances an echo system. I want to make sure we mean deprecating the concept all together, not just replacing it with something else

Reply:
If the are no objections I'll close this thread.